File: solutions/leetcode_368.py
from functools import cache


class Solution:
    def largestDivisibleSubset(self, nums: list[int]) -> list[int]:
        nums.sort()

        @cache
        def helper(i, prev):
            if i == len(nums):
                return []

            result = []
            for j in range(i, len(nums)):
                if nums[j] % prev == 0:
                    local_result = [nums[j]] + helper(j + 1, nums[j])
                    if len(local_result) > len(result):
                        result = local_result
            return result


        result = []
        for i in range(len(nums)):
            local_result = [nums[i]] + helper(i + 1, nums[i])
            if len(local_result) > len(result):
                result = local_result

        return result


if __name__ == '__main__':
    given_nums = [8, 4, 2, 1]
    expected = [1, 2, 4, 8]

    solution = Solution()
    result = solution.largestDivisibleSubset(given_nums)

    assert result == expected

File: solutions/leetcode_368.swift
class Solution {
    func largestDivisibleSubset(_ nums: [Int]) -> [Int] {
        var nums = nums.sorted()
        var dp = [[Int]]()
        for num in nums {
            dp.append([num])
        }

        for i in 0..<(nums.count - 1) {
            for j in (i + 1)..<nums.count {
                if nums[j] % nums[i] == 0 {
                    if dp[i].count + 1 > dp[j].count {
                        dp[j] = [nums[j]] + dp[i]
                    }
                }
            }
        }

        var index = 0
        for i in 1..<dp.count {
            if dp[i].count > dp[index].count {
                index = i
            }
        }

        return dp[index]
    }
}
